generics/straightDriveBase/drive.cpp

```diff
@@ -0,0 +1,312 @@
+        //lcd::set_text(0, "speed > 0: " + to_string(speed) + " bias: " + to_string(LRBiasLow));
+    }
+    //backwards biases
+    else if (speed < -maxSpeed)
+    {
+        speed = -maxSpeed;
+        rawSetMotors(speed, LRBiasHighBack);
+        //lcd::set_text(0, "speed < -maxSpeed: " + to_string(speed) + " bias: " + to_string(LRBiasHighBack));
+    }
+    else if (speed < -midSpeed)
+    {
+        double highPercent = (-speed - midSpeed) / (maxSpeed - midSpeed);
+        double lowPercent = (maxSpeed + speed) / (maxSpeed - midSpeed);
+
+        rawSetMotors(speed, LRBiasLowBack * lowPercent + LRBiasHighBack * highPercent);
+        //lcd::set_text(0, "speed < -midSpeed: " + to_string(speed) + " bias: "
+        //        + to_string(LRBiasLow * lowPercent + LRBiasHigh * highPercent));
+    }
+    else //if (speed < 0)
+    {
+        rawSetMotors(speed, LRBiasLowBack);
+        //lcd::set_text(0, "speed < 0: " + to_string(speed) + " bias: " + to_string(LRBiasLowBack));
+    }
+}
+
+//tqke in a vector of motors, and call the move relative function for all of them with a given distance and speed
+void FourWheelDrive::setMotorsRelative(MinesMotorGroup * motors, double distance, double speed)
+{
+  motors->spinFor(distance, ROT_UNIT, speed, VEL_UNIT);
+}
+
+void FourWheelDrive::setMotorsRelative(double distance, double speed)
+{
+  setMotorsRelative(rightMotors, distance, speed);
+  setMotorsRelative(leftMotors, distance, speed);
+}
+
+void FourWheelDrive::setZeroPosition()
+{
+  leftMotors->resetPosition();
+  rightMotors->resetPosition();
+}
+
+double FourWheelDrive::getPosition(MinesMotorGroup * motors)
+{
+    return motors->position(ROT_UNIT);
+}
+
+double FourWheelDrive::getAllPosition()
+{
+    return (getPosition(rightMotors) + getPosition(leftMotors)) / 2.0;
+}
+
+double FourWheelDrive::getSpeed(MinesMotorGroup * motors)
+{
+  return motors->velocity(VEL_UNIT);
+}
+
+double FourWheelDrive::getAllSpeed()
+{
+    return (getSpeed(rightMotors) + getSpeed(leftMotors)) / 2.0;
+}
+
+//TODO port a version to MinesMotorGroup
+void FourWheelDrive::accelerate(double targetSpeed)
+{
+	const double TOLERANCE = 0.5;
+
+	double currSpeed = getAllSpeed();
+	if(currSpeed < targetSpeed)
+	{
+		while(currSpeed < targetSpeed - TOLERANCE)
+		{
+			currSpeed += maxAccelerationForward;
+			setMotors(currSpeed);
+			task::sleep(LOOP_DELAY);
+		}
+	}
+	else
+	{
+		while(currSpeed > targetSpeed + TOLERANCE)
+		{
+			currSpeed -= maxAccelerationBackward;
+			setMotors(currSpeed);
+			task::sleep(LOOP_DELAY);
+		}
+	}
+}
+
+void FourWheelDrive::driveTilesPID(float numTiles, float desiredSpeed)
+{
+    float INTEGRATOR_MAX_MAGNITUDE = 1000;
+    float DELTA_T = LOOP_DELAY / 1000.0;
+    const int STOP_LOOPS = 20;
+    const float TILE_TOLERANCE = 0.02;
+    const float DESIRED_SPEED = 70;
+    // 4 Inches wheels, 600RPM motors, measured 222.22 ticks/rotation
+    const double TICKS_PER_TILE = 1333.3;
+    float currentDistance = 0;
```

Comment:
FourWheelDrive::driveTilesPID() and FourWheelDrive::turnDegreesAbsolutePID() have the same variables. This could be a header variable instead. The variables are also all caps, so maybe these should be global constants.